// @nestjs/microservices/external/kafka-options.interface.js
"use strict";
/// <reference types="node" />
Object.defineProperty(exports, "__esModule", { value: true });
var ResourceType;
(function (ResourceType) {
    ResourceType[ResourceType["UNKNOWN"] = 0] = "UNKNOWN";
    ResourceType[ResourceType["ANY"] = 1] = "ANY";
    ResourceType[ResourceType["TOPIC"] = 2] = "TOPIC";
    ResourceType[ResourceType["GROUP"] = 3] = "GROUP";
    ResourceType[ResourceType["CLUSTER"] = 4] = "CLUSTER";
    ResourceType[ResourceType["TRANSACTIONAL_ID"] = 5] = "TRANSACTIONAL_ID";
    ResourceType[ResourceType["DELEGATION_TOKEN"] = 6] = "DELEGATION_TOKEN";
})(ResourceType = exports.ResourceType || (exports.ResourceType = {}));
var logLevel;
(function (logLevel) {
    logLevel[logLevel["NOTHING"] = 0] = "NOTHING";
    logLevel[logLevel["ERROR"] = 1] = "ERROR";
    logLevel[logLevel["WARN"] = 2] = "WARN";
    logLevel[logLevel["INFO"] = 4] = "INFO";
    logLevel[logLevel["DEBUG"] = 5] = "DEBUG";
})(logLevel = exports.logLevel || (exports.logLevel = {}));
var CompressionTypes;
(function (CompressionTypes) {
    CompressionTypes[CompressionTypes["None"] = 0] = "None";
    CompressionTypes[CompressionTypes["GZIP"] = 1] = "GZIP";
    CompressionTypes[CompressionTypes["Snappy"] = 2] = "Snappy";
    CompressionTypes[CompressionTypes["LZ4"] = 3] = "LZ4";
    CompressionTypes[CompressionTypes["ZSTD"] = 4] = "ZSTD";
})(CompressionTypes = exports.CompressionTypes || (exports.CompressionTypes = {}));
