// @nestjs/terminus/dist/terminus.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var TerminusModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("@nestjs/common");
const terminus_core_module_1 = require("./terminus-core.module");
const diskusage_lib_provider_1 = require("./health-indicator/disk/diskusage-lib.provider");
const health_indicators_provider_1 = require("./health-indicator/health-indicators.provider");
const health_check_1 = require("./health-check");
const health_check_executor_service_1 = require("./health-check/health-check-executor.service");
/**
 * The Terminus module integrates health checks
 * and graceful shutdowns in your Nest application
 *
 * @publicApi
 */
let TerminusModule = TerminusModule_1 = class TerminusModule {
    /**
     * Bootstraps the Terminus Module synchronously
     * @param options The options for the Terminus Module
     * @deprecated
     */
    static forRoot(options) {
        return {
            module: TerminusModule_1,
            imports: [terminus_core_module_1.TerminusCoreModule.forRoot(options)],
        };
    }
    /**
     * Bootstrap the Terminus Module asynchronously
     * @param options The options for the Terminus module
     */
    static forRootAsync(options) {
        return {
            module: TerminusModule_1,
            imports: [terminus_core_module_1.TerminusCoreModule.forRootAsync(options)],
        };
    }
};
TerminusModule = TerminusModule_1 = __decorate([
    common_1.Module({
        imports: [common_1.HttpModule],
        providers: [
            diskusage_lib_provider_1.DiskusageLibProvider,
            health_check_executor_service_1.HealthCheckExecutor,
            health_check_1.HealthCheckService,
            ...health_indicators_provider_1.HEALTH_INDICATORS,
        ],
        exports: [health_check_1.HealthCheckService, ...health_indicators_provider_1.HEALTH_INDICATORS],
    })
], TerminusModule);
exports.TerminusModule = TerminusModule;
